grades/api_test.py

```diff
@@ -0,0 +1,237 @@
+        assert sorted([user.pk for user in
+                       api.get_users_final_grade_freeze(self.course_run1)]) == sorted([self.user.pk, other_user.pk])
+
+        # add the user to the FinalGrade model as in progress
+        fg_status = FinalGrade.objects.create(
+            user=other_user, course_run=self.course_run1, status=FinalGradeStatus.PENDING, grade=0.0)
+        assert sorted([user.pk for user in
+                       api.get_users_final_grade_freeze(self.course_run1)]) == sorted([self.user.pk, other_user.pk])
+
+        # change the final grade status to complete
+        fg_status.status = FinalGradeStatus.COMPLETE
+        fg_status.save()
+        assert [user.pk for user in api.get_users_final_grade_freeze(self.course_run1)] == [self.user.pk]
+
+    @patch('grades.api.get_final_grade', autospec=True)
+    @patch('grades.api._refresh_cache_final_grade', autospec=True)
+    def test_freeze_user_final_grade_error1(self, mock_refr, mock_get_fg):
+        """
+        Test for freeze_user_final_grade function in case of problems with can_freeze_grades
+        """
+        # case not ready to be frozen
+        api.freeze_user_final_grade(self.user, self.course_run3)
+        assert mock_refr.called is False
+        assert mock_get_fg.called is False
+        assert FinalGrade.objects.filter(user=self.user, course_run=self.course_run3).exists() is False
+
+        # case without freeze date
+        with self.assertRaises(ImproperlyConfigured):
+            api.freeze_user_final_grade(self.user, self.course_run2)
+        assert mock_refr.called is False
+        assert mock_get_fg.called is False
+        assert FinalGrade.objects.filter(user=self.user, course_run=self.course_run2).exists() is False
+
+    @patch('grades.api.get_final_grade', autospec=True)
+    @patch('grades.api._refresh_cache_final_grade', autospec=True)
+    def test_freeze_user_final_grade_error2(self, mock_refr, mock_get_fg):
+        """
+        Test for freeze_user_final_grade function in case of problems with refresh of cache
+        """
+        mock_refr.side_effect = AttributeError
+        api.freeze_user_final_grade(self.user, self.course_run1)
+        assert mock_get_fg.called is False
+        mock_refr.assert_called_once_with(self.user)
+        assert FinalGrade.objects.filter(user=self.user, course_run=self.course_run1).exists() is False
+
+    @patch('grades.api.get_final_grade', autospec=True)
+    @patch('grades.api._refresh_cache_final_grade', autospec=True)
+    def test_freeze_user_final_grade_error3(self, mock_refr, mock_get_fg):
+        """
+        Test for freeze_user_final_grade function in case of problems with getting the final grade
+        """
+        mock_get_fg.side_effect = AttributeError
+        api.freeze_user_final_grade(self.user, self.course_run1)
+        mock_refr.assert_called_once_with(self.user)
+        mock_get_fg.assert_called_once_with(self.user, self.course_run1)
```

Comment:
The tested function is structured in a way to never fail (or at least it should not to). The only way to see if the function stops the execution at the expected points is to check if the some calls are or aren't executed. I am open to other solutions to accomplish the same.